### ISSUE TITLE
json request for certain media types fail

### ISSUE THREAD

**Issue by iherman** [aio-libs/aiohttp#2206]
## Long story short

The `resp.json()` call reports a bug if the return type is something like `application/XXX+json` (as opposed to `application/json`). However, `application/XXX+json` is a perfectly fine media type, "extending" the _semantics_ (not the syntax) of the core json syntax, ie, should be parsed as bona fide json...

## Your environment

OS X 10.11.6, Python 3.6.2

**Comment by jscholes** [comment]
You can pass a custom content_type parameter to the ClientResponse.json method, although I agree that this isn't perhaps the most sensible API.  I just switched some code from using Requests to using this library and while Requests could decode the JSON without a problem, aiohttp reported an error which will trip some people up.

**Comment by asvetlov** [comment]
Related to #2174

**Comment by asvetlov** [comment]
Extending the list of accepted json content types is fair easy change.
Please propose a Pull Request. `application/XXX+json` is not acceptable but I pretty sure you could grab a list of wide spread json types from `requests` library.

**Comment by asvetlov** [comment]
use `content_type=None` for disabling the check

**Comment by asvetlov** [comment]
What to do?
On one hand we want to prevent processing non-JSON data in `.json()` method.
On other we don't want to be too strict.

Maybe you can pass a custom content type when fetching a data from your server?
Or create a custom json reading function on top of `resp.read()`?

**Comment by iherman** [comment]
@asvetlov,

The fundamental fact is that data coming with `XXX+json` **is** JSON. From the point of view of aiohttp that all that matters. A properly registered `XXX+json` media type means that the content _is_ JSON but it may have some additional syntactic or structural constraint _within the framework of JSON_. This may mean imposing a particular JSON Schema, or some more complicated specification on top of what JSON means, or may be no extra syntactic restriction whatsoever. 

To take a specific example. `application/ld+json` is a bona fide media type defined through the structural syntax; the content is defined via the [JSON-LD Recommendation](https://www.w3.org/TR/json-ld/). That specification adds some additional semantics to individual terms, requires some specific structure to the JSON data, but the essential point is that a JSON-LD content is a perfectly fine JSON content. (This is probably one of the most complex examples. Others may be a few paragraphs only.)

In my view, this means that the handling of the `XXX+json` (or `XXX+xml`, etc.) is simple, as far as aiohttp goes: strip the `XXX+` string altogether (the `+` character is not allowed elsewhere in a media type name), and handle the content as `json` (or `xml`). You are done. All other processing is in the application domain.

(You will find an abundance of such media types on the [media type registry](https://www.iana.org/assignments/media-types/media-types.xhtml).)

**Comment by asvetlov** [comment]
You would say that any `application/xxx+json` is a valid JSON type.
Or parsing all JSON registered types from the media registry is better?

**Comment by webknjaz** [comment]
I'd say both statements are true.

**Comment by webknjaz** [comment]
By sending `application/xxx+json` client explicitly tells server to expect valid json payload and vice versa

**Comment by iherman** [comment]
@asvetlov : parsing the json types from the media registry is of course nicer and better but... it is not that obvious. Would aiohttp check that registry any time the "read" function is called? Would set up some sort of a local cache for the registry that has a finite lifetime? The registry is a changing target after all...

I believe just saying that "any `application/xxx+json` is a valid JSON type" is good enough.

**Comment by asvetlov** [comment]
Reopen the issue.
@iherman would you provide a Pull Request with the first option (accepting `application/xxx+json`)?

**Comment by iherman** [comment]
@asvetlov : I have no idea when I could do that; I have never looked into the internal of the aiohttp code, ie, it would take me quite a while to familiarize myself with it to produce a proper PR. I can try at some point unless somebody beats me into it, but it may take a while:-(

**Comment by iherman** [comment]
@asvetlov, I have looked at the code and I _think_ I have found the few changes that are required. However, I simply will not have the time to make a proper pull request, ie, running all the necessary tests on the code overall. It would require a major setup on my machine, understanding how the test harness works, etc. Sorry, but that will not work.

I will look at the code again tomorrow just to be sure, and I could then forward you (tomorrow or on Wednesday) the file where I have made some changes: `client_reqrep.py`. Would that work?
